src/workers/viewportFeatures.worker.js

```diff
@@ -0,0 +1,87 @@
+import { viewportFeatures } from '../api/viewportFeaturesBinary';
+import { Methods } from './workerMethods';
+import { aggregationFunctions } from '../widgets/operations/aggregation/values';
+import { buildFeatureFilter } from '../api/Filter';
+import { histogram } from '../widgets/operations/histogram';
+import { groupValuesByColumn } from '../widgets/operations/groupby';
+
+let currentViewportFeatures;
+
+onmessage = ({ data: { method, ...params } }) => {
+  switch (method) {
+    case Methods.VIEWPORT_FEATURES:
+      getViewportFeatures(params);
+      break;
+    case Methods.VIEWPORT_FEATURES_FORMULA:
+      getFormula(params);
+      break;
+    case Methods.VIEWPORT_FEATURES_HISTOGRAM:
+      getHistogram(params);
+      break;
+    case Methods.VIEWPORT_FEATURES_CATEGORY:
+      getCategories(params);
+      break;
+    default:
+      throw new Error('Invalid worker method');
+  }
+};
+
+function getViewportFeatures({ tiles, viewport, uniqueIdProperty }) {
+  currentViewportFeatures = viewportFeatures({
+    tiles,
+    viewport,
+    uniqueIdProperty
+  });
+  postMessage({ result: true });
+}
+
+function getFormula({ filters, operation, column }) {
+  let result = [{ value: null }];
+
+  if (currentViewportFeatures) {
+    const targetOperation = aggregationFunctions[operation];
+
+    const filteredFeatures = !Object.keys(currentViewportFeatures).length
+      ? currentViewportFeatures
+      : currentViewportFeatures.filter(buildFeatureFilter({ filters }));
```

Comment:
These 3 lines could probably be extracted into a method, as it looks they are repeated on each method

Reply:
I copied the functions from models and I did not stop to see the code. I have added the `getFilteredFeatures ` function